src/lsm/compaction.zig

```diff
@@ -311,293 +323,256 @@ pub fn CompactionType(
+            assert(compaction.status == .processing);
+            assert(compaction.callback != null);
+            assert(compaction.io_pending == 0);
 
-            assert(!block_write.ready);
-            block_write.ready = true;
-        }
+            // Ensure merging is truly finished.
+            assert(compaction.merge_status == .processing);
+            assert(compaction.merge_iterator.empty());
+            assert(!compaction.data.ready);
+            assert(!compaction.filter.ready);
+            assert(!compaction.index.ready);
 
-        fn assert_read_iterators_empty(compaction: Compaction) void {
+            // Double check the iterators are finished as well.
             assert(compaction.iterator_a.buffered_all_values());
             assert(compaction.iterator_a.peek() == null);
-
             assert(compaction.iterator_b.buffered_all_values());
             assert(compaction.iterator_b.peek() == null);
-        }
 
-        fn stream_peek(compaction: *Compaction, stream_id: u32) ?Key {
-            assert(stream_id <= 1);
+            // Apply any pending manifest changes queued up during the merge.
+            compaction.apply_manifest_updates(&compaction.update_level_b);
+            compaction.apply_manifest_updates(&compaction.insert_level_b);
 
-            if (stream_id == 0) {
-                return compaction.iterator_a.peek();
-            } else {
-                return compaction.iterator_b.peek();
+            // Remove the level_a table if it was provided given it's now been merged into level_b.
+            // TODO: Release the grid blocks associated with level_a as well
+            if (compaction.level_b != 0) {
+                const level_a = compaction.level_b - 1;
+
+                if (compaction.remove_level_a) |*level_a_table| {
+                    const tables = @as(*[1]TableInfo, level_a_table);
+                    compaction.manifest.update_tables(level_a, compaction.snapshot, tables);
+                }
             }
+
+            // Finally, mark Compaction as officially complete and ready to be reset().
+            compaction.merge_status = .done;
+            compaction.status = .done;
         }
 
-        fn stream_pop(compaction: *Compaction, stream_id: u32) Value {
-            assert(stream_id <= 1);
+        pub fn reset(compaction: *Compaction) void {
+            assert(compaction.status == .done);
+            assert(compaction.callback == null);
+            assert(compaction.io_pending == 0);
+            assert(compaction.merge_status == .done);
 
-            if (stream_id == 0) {
-                return compaction.iterator_a.pop();
-            } else {
-                return compaction.iterator_b.pop();
-            }
-        }
+            // Double check that manifest updates have been applied.
+            assert(compaction.update_level_b.drain().len == 0);
+            assert(compaction.insert_level_b.drain().len == 0);
 
```

Comment:
Can we also re-assert that the iterators are all done?

Reply:
This happens above